sim/midas/src/main/cc/simif_token_hashers.cc

```diff
@@ -0,0 +1,222 @@
+#include "simif_token_hashers.h"
+#include "simif.h"
+
+#include <iostream>
+#include <sstream>
+#include <fstream>
+
+/**
+ * The constructor for simif_token_hashers_t.
+ *
+ * @param [in] p A pointer to the parent instance of simif_t
+ */
+simif_token_hashers_t::simif_token_hashers_t(simif_t *p) : parent(p) {
+
+#ifdef TOKENHASHMASTER_0_PRESENT
+  TOKENHASHMASTER_0_substruct_create;
+  trigger0 = TOKENHASHMASTER_0_substruct->triggerDelay0_TokenHashMaster;
+  trigger1 = TOKENHASHMASTER_0_substruct->triggerDelay1_TokenHashMaster;
+  period0 = TOKENHASHMASTER_0_substruct->triggerPeriod0_TokenHashMaster;
+  period1 = TOKENHASHMASTER_0_substruct->triggerPeriod1_TokenHashMaster;
+  free(TOKENHASHMASTER_0_substruct);
+
+  cnt = TOKENHASH_COUNT;
+  bridge_names.assign(TOKENHASH_BRIDGENAMES, TOKENHASH_BRIDGENAMES + cnt);
+  names.assign(TOKENHASH_NAMES, TOKENHASH_NAMES + cnt);
+  outputs.assign(TOKENHASH_OUTPUTS, TOKENHASH_OUTPUTS + cnt);
+  queue_heads.assign(TOKENHASH_QUEUEHEADS, TOKENHASH_QUEUEHEADS + cnt);
+  queue_occupancies.assign(TOKENHASH_QUEUEOCCUPANCIES,
+                           TOKENHASH_QUEUEOCCUPANCIES + cnt);
+  tokencounts0.assign(TOKENHASH_TOKENCOUNTS0, TOKENHASH_TOKENCOUNTS0 + cnt);
+  tokencounts1.assign(TOKENHASH_TOKENCOUNTS1, TOKENHASH_TOKENCOUNTS1 + cnt);
+#endif
+}
+
+/**
+ * Print debug info about the MMIO internals
+ */
+void simif_token_hashers_t::info() {
```

Comment:
But for debugging methods I'd try to introduce something along the lines of:

```
void info(std::ostream &os = std::cout)
```

If we want to dump stuff at some point to a file or to stderr instead, we should be able to. LLVM/MLIR/CIRCT also often parameterize by a stream and default to a standard one.